### PR TITLE
Aws console logs

### PR DESCRIPTION
**Pull request by r4f4**
Depends on https://github.com/openshift/installer/pull/5697

https://issues.redhat.com/browse/CORS-1945

### CODE REVIEW

**Comment by r4f4** [comment]
/hold
until I can do some proper testing

**Comment by patrickdillon** [comment]
For the combined tar file, can we flatten to a dir structure like this:
```
log-bundle-20220511112307
├──  bootstrap
├── control-plane
├──  failed-units.txt
├──  gather.log
├──  rendered-assets
├──  resources
├──  unit-status
└── serial
 
```

Currently the structure is like this:
```
log-bundle-20220511112307
├── log-bundle-20220511112307
│   ├── bootstrap
│   ├── control-plane
│   ├── failed-units.txt
│   ├── gather.log
│   ├── rendered-assets
│   ├── resources
│   └── unit-status
└── serial

```
In other words, can we move everything in `log-bundle-20220511112307/log-bundle-20220511112307`
 up to `log-bundle-20220511112307` and then delete `log-bundle-20220511112307/log-bundle-20220511112307`

File: cmd/openshift-install/gather.go [hunk 1]
```diff
@@ -142,6 +142,25 @@ func runGatherBootstrapCmd(directory string) (string, error) {
...
+	gatherID := time.Now().Format("20060102150405")
+
+	serialLogBundle := filepath.Join(directory, fmt.Sprintf("serial-log-bundle-%s.tar.gz", gatherID))
+	serialLogBundlePath, err := filepath.Abs(serialLogBundle)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to stat log file")
+	}
+
+	logrus.Info("Pulling VM console logs")
+	consoleGather, err := serialgather.New(logrus.StandardLogger(), serialLogBundle, bootstrap, masters, directory)
```
Comment (patrickdillon) [inline cmd/openshift-install/gather.go:154]: should this be `serialLogBundle` as it is or `serialLogBundlePath`?

File: pkg/gather/aws/aws.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+	if len(files) > 0 {
+		err := gather.CreateArchive(files, g.serialLogBundle)
+		if err != nil {
+			errs = append(errs, errors.Wrap(err, "failed to create archive"))
+		}
+	}
+
+	if err := gather.DeleteArchiveDirectory(filePathDir); err != nil {
+		// Note: cleanup is best effort, it shouldn't fail the gather
+		g.logger.Debugf("failed to remove archive directory: %v", err)
```
Comment (patrickdillon) [inline pkg/gather/aws/aws.go:135]: nit: unlike error messages, log messages are capitalized
```suggestion
		g.logger.Debugf("Failed to remove archive directory: %v", err)
```

File: pkg/gather/aws/aws.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+// findEC2Instances returns the EC2 instances with tags that satisfy the filters.
+func (g *Gather) findEC2Instances(ctx context.Context, ec2Client *ec2.EC2) ([]*ec2.Instance, error) {
+	if ec2Client.Config.Region == nil {
+		return nil, errors.New("EC2 client does not have region configured")
+	}
+
+	var instances []*ec2.Instance
+	for _, filter := range g.filters {
+		g.logger.Debugf("search for matching instances by tag in %s matching %#+v", *ec2Client.Config.Region, filter)
```
Comment (patrickdillon) [inline pkg/gather/aws/aws.go:149]:
```suggestion
		g.logger.Debugf("Search for matching instances by tag in %s matching %#+v", *ec2Client.Config.Region, filter)
```

File: pkg/gather/aws/aws.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+			err = errors.Wrap(err, "get ec2 instances")
+			return instances, err
+		}
+	}
+
+	return instances, nil
+}
+
+func (g *Gather) downloadConsoleOutput(ctx context.Context, ec2Client *ec2.EC2, instance *ec2.Instance, filePathDir string) (string, error) {
+	logger := g.logger.WithField("instance", *instance.InstanceId)
```
Comment (patrickdillon) [inline pkg/gather/aws/aws.go:186]:
```suggestion
	logger := g.logger.WithField("Instance", *instance.InstanceId)
```

File: pkg/gather/aws/aws.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+	}
+
+	instanceName := *result.InstanceId
+	for _, tags := range instance.Tags {
+		if strings.EqualFold(*tags.Key, "Name") {
+			instanceName = *tags.Value
+		}
+	}
+
+	logger.Debugf("attemping to download console logs for %s", instanceName)
```
Comment (patrickdillon) [inline pkg/gather/aws/aws.go:207]:
```suggestion
	logger.Debugf("Attemping to download console logs for %s", instanceName)
```

File: pkg/gather/aws/aws.go [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+	filePath, err := g.saveToFile(instanceName, *result.Output, filePathDir)
+	if err != nil {
+		return "", err
+	}
+
```
Comment (patrickdillon) [inline pkg/gather/aws/aws.go:212]:
```suggestion
logger.Debug("Download complete")

```

**Review by patrickdillon — COMMENTED** [review]
These are mostly nits or small changes. I think this is in good shape.

File: cmd/openshift-install/gather.go [hunk 7]
```diff
@@ -151,19 +170,28 @@ func gatherBootstrap(bootstrap string, port int, masters []string, directory str
...
-	path, err := filepath.Abs(file)
+
+	clusterLogBundlePath, err := filepath.Abs(file)
 	if err != nil {
 		return "", errors.Wrap(err, "failed to stat log file")
 	}
-	return path, nil
+
+	logBundlePath := filepath.Join(filepath.Dir(clusterLogBundlePath), fmt.Sprintf("log-bundle-%s.tar.gz", gatherID))
+	archives := map[string]string{serialLogBundlePath: "serial", clusterLogBundlePath: ""}
```
[the remark this answers is not in the document]
Comment (r4f4) [reply cmd/openshift-install/gather.go:188]: If the `serialLogBundle` archive doesn't not exist or if it contains no files, there will be no `serial` directory in the combined bundle.

File: pkg/gather/aws/aws.go [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+	instanceName := *result.InstanceId
+	for _, tags := range instance.Tags {
+		if strings.EqualFold(*tags.Key, "Name") {
+			instanceName = *tags.Value
+		}
+	}
+
+	logger.Debugf("attemping to download console logs for %s", instanceName)
+	filePath, err := g.saveToFile(instanceName, *result.Output, filePathDir)
```
Comment (r4f4) [inline pkg/gather/aws/aws.go:208]: It just happened to me now that `result.Output` is `nil`. I'll fix this so we don't dereference a `nil` pointer.

File: cmd/openshift-install/gather.go [hunk 9]
```diff
@@ -151,19 +170,28 @@ func gatherBootstrap(bootstrap string, port int, masters []string, directory str
...
+	err = serialgather.CombineArchives(logBundlePath, archives)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to combine archives")
+	}
+
```
[the remark this answers is not in the document]
Comment (r4f4) [reply cmd/openshift-install/gather.go:192]: Done

**Comment by r4f4** [comment]
I'll squash the commits and drop the DO NOT MERGE commit when the tests show positive results.

**Pull request merged**